### PR TITLE
Add model utility functions

### PR DESCRIPTION
**Pull request by sabaini**
Add function to block until file on units matches a regular expression,
and a function to retrieve remote file contents

### CODE REVIEW

File: zaza/model.py [hunk 1]
```diff
@@ -1316,6 +1322,32 @@ async def _block():
...
+        run_func = run_on_leader
+    else:
+        target = unit_name
+        run_func = run_on_unit
+    cmd = "cat {}".format(path)
+    result = run_func(target, cmd, timeout=timeout)
+    err = result.get("Stderr")
+    if err:
+        raise RemoteFileError(err)
+    return result.get("Stdout")
```
Comment (ajkavanagh) [inline zaza/model.py:1348]: Isn't there a Code that indicates an error too?  The `RemoteFileError` might be more useful if it also contained the error code from the cat?

Comment (ajkavanagh) [reply zaza/model.py:1348]: I'm a bit confused as to why there are sync and async functions in this PR?  I'm a bit worried that `file_contents()` is being used in an async test function?  It just seems a bit strange.  Please could you show how these are being used?  Thanks.

Comment (sabaini) [reply zaza/model.py:1348]: > Isn't there a Code that indicates an error too? The `RemoteFileError` might be more useful if it also contained the error code from the cat?

The `cat` exit codes aren't terribly interesting -- they're 0 on success and 1 on failure. Do you think it's worthwhile including them?

Comment (sabaini) [reply zaza/model.py:1348]: > I'm a bit confused as to why there are sync and async functions in this PR? I'm a bit worried that `file_contents()` is being used in an async test function? It just seems a bit strange. Please could you show how these are being used? Thanks.

I've provided an async function for blocking on regexp as well as I thought there might be a usecase for expensive regexp matching. I didn't think there would be much use for an async `file_contents()`

Do you want me to provide functional testing for those?

Comment (ajkavanagh) [reply zaza/model.py:1348]: > > Isn't there a Code that indicates an error too? The `RemoteFileError` might be more useful if it also contained the error code from the cat?
> 
> The `cat` exit codes aren't terribly interesting -- they're 0 on success and 1 on failure. Do you think it's worthwhile including them?

Actually, no, in that case, it's not terribly useful.  You made the right call on that.

Comment (ajkavanagh) [reply zaza/model.py:1348]: > > I'm a bit confused as to why there are sync and async functions in this PR? I'm a bit worried that `file_contents()` is being used in an async test function? It just seems a bit strange. Please could you show how these are being used? Thanks.
> 
> I've provided an async function for blocking on regexp as well as I thought there might be a usecase for expensive regexp matching. I didn't think there would be much use for an async `file_contents()`
> 
> Do you want me to provide functional testing for those?

If I understand correctly, there isn't actually a use-case for `file_contents()`?  i.e. there isn't a test in zaza-openstack-tests that uses it?  In that case, there is no need to add it here.  This is affectionately termed 'gold-plating' ... the code has to be maintained, but nothing is using it, and it's been added 'just-in-case'.

Is this *also* the case for the other functions in this PR?

If I've misunderstood, my apologies!

Comment (sabaini) [reply zaza/model.py:1348]: I meant to say that I didn't see a use case for an _async_ version of `file_contents()` -- only for the synchronous version. If so desired I could also create an async version of `file_contents()` though.

The functions added here indeed come from a concrete use case. I've used their predecessor versions in the launchpad MR [here](https://code.launchpad.net/~peter-sabaini/charm-juju-lint/+git/charm-juju-lint/+merge/394000). 

In short: 

- the `file_contents()` is used to get values from a test dependency which are then used to configure the system. The test dependency is a juju controller created within the test model, from where I need to retrieve the generated model parameters to then feed them into the Juju lint charm.

- `block_until_file_matches_re()` is used to test for SUT output which is better described as a pattern, as there are some varying bits in there which aren't relevant for the test. The test output being from the auto-lint program where I'd like to match on the output keywords but not on the generated hostname, time stamps etc.

I hope this clarifies? If not feel free to also ping me on IRC

**Comment by ajkavanagh** [comment]
I'm generally negative in terms of merging this PR purely because no charm test (that uses zaza) is actually currently using it and thus it runs the risk of bit rot and is essentially a bit of gold plating.  That's not to say that it is not good code, just that it's perhaps not yet required.  However, I won't block anyone else from merging it, if they think it is useful to just put in the library.

**Comment by sabaini** [comment]
Hey, 

as discussed elsewhere I've now refactored a client of Zaza to make use of the proposed utilities [here](https://code.launchpad.net/~peter-sabaini/charm-juju-lint/+git/charm-juju-lint/+merge/394000), in the hope to reduce code duplication for this and similar charms.

cheers!

**Review by ajkavanagh — APPROVED** [review]
LGTM

**Pull request merged**